### ISSUE TITLE
Dynamic Drop

### ISSUE THREAD

**Issue by Gankro** [Manishearth/rust-clippy#431]
Some people are die-hard fans of static drop semantics as a way to make sure both branches do basically the same thing. Would it be possible to have a lint against using dynamic drop?

**Comment by Manishearth** [comment]
In other words, just lint against `let x;`?

**Comment by Gankro** [comment]
Simple example:

```rust
let x = vec![1, 2 , 3];
if foo {
  // move x out conditionally
  mem::drop(x);
}

// ... stuff
// x maybe dropped here?
```

The vec will dynamically track what branch is taken with drop flags to know if it should be dropped or not. Static drop would require that an else is given that also moves `x`:

```rust
let x = vec![1, 2 , 3];
if foo {
  mem::drop(x);
} else {
  mem::drop(x);
}
```

**Comment by Manishearth** [comment]
That's still static drop. Or, well, it can be. I thought Felix had plans to make everything except deferred-let a static drop.

**Comment by Gankro** [comment]
That's new information to me. You might be thinking of "eager drop", which would be a breaking change?

https://github.com/mmun/rfcs-1/blob/master/text/0320-nonzeroing-dynamic-drop.md

https://github.com/rkjnsn/rfcs/blob/allow-eager-drops/active/0000-allow-eager-drops.md

CC @pnkfelix

**Comment by Manishearth** [comment]
Oh, wait, yes, eager drop is necessary for this to work.

But I had heard of plans of making everything except deferred let static. Not sure what it was.

**Comment by Gankro** [comment]
You could probably desugar all conditional moves to have the `else` branch move into a deferred let in e.g. the MIR, I guess.
